discovery-provider/src/queries/get_premium_track_signatures.py

```diff
@@ -0,0 +1,337 @@
+    # using all user eth wallets, and generate signatures for corresponding tracks.
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        future_to_contract_address_map = {
+            executor.submit(
+                _does_user_own_erc721_nft_collection, contract_address, user_eth_wallets
+            ): contract_address
+            for contract_address in list(erc721_collection_track_map.keys())
+        }
+        for future in concurrent.futures.as_completed(future_to_contract_address_map):
+            contract_address = future_to_contract_address_map[future]
+            try:
+                # Generate premium content signatures for tracks whose
+                # nft collection is owned by the user.
+                if future.result():
+                    for track_id in erc721_collection_track_map[contract_address]:
+                        track_signature_map[track_id] = get_premium_content_signature(
+                            {
+                                "id": track_id,
+                                "type": "track",
+                                "user_wallet": user_wallet,
+                            }
+                        )
+            except Exception as e:
+                logger.error(
+                    f"Could not future result for erc721 contract_address {contract_address}. Error: {e}"
+                )
+
+    # Handle premium tracks gated by ERC1155 nft collections
+    erc1155_gated_tracks = list(
+        filter(
+            lambda track: track.premium_conditions["nft_collection"]["standard"]  # type: ignore
+            == "ERC1155",
+            tracks,
+        )
+    )
+
+    # Build a map of ERC1155 collection address -> track ids
+    # so that only one chain call will be made for premium tracks
+    # that share the same nft collection gate.
+    # Also build a map of ERC1155 collection address -> nft token ids
+    # so that the balanceOf contract function can be used to check
+    # batch ownership of ERC1155 nfts for a given collection.
+    erc1155_collection_track_map = defaultdict(list)
+    contract_address_token_id_map: Dict[str, Set[int]] = defaultdict(set)
+    for track in erc1155_gated_tracks:
+        contract_address = Web3.toChecksumAddress(
+            track.premium_conditions["nft_collection"]["address"]  # type: ignore
+        )
+        erc1155_collection_track_map[contract_address].append(track.track_id)
+        track_token_id_set = set(map(int, track_token_id_map[track.track_id]))
+        contract_address_token_id_map[contract_address] = contract_address_token_id_map[
+            contract_address
+        ].union(track_token_id_set)
+
+    # Check ownership of nfts from erc1155 collections from given contract addresses,
+    # using all user eth wallets, and generate signatures for corresponding tracks.
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
could avoid creating a new threadpool executor by nesting both checks in the same `with` block

Reply:
ah. nice